Light Tiles use a Radial Menu and other tweaks

RaveRadbury opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Light tiles have a radial menu now. Also, I've added orange to light tiles. Light tiles now default to blue when placed. You need to use a multitool on them to get them to change color. There's now a second color cycle setting. I've pulled white out of every color cycling tile type and ROYGBIV sorted them.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
https://file.house/MWGW.mp4 Side-by-side comparison, new on the right, old on the left.

Players put all of this effort into very specific tile designs and then well meaning players ruin them with simple misclicks. Not anymore! Color cycling is improved. The full color spectrum is represented.

![image](https://user-images.githubusercontent.com/3204033/80306485-7c551f00-8778-11ea-9720-832773b3e01e.png)
I was able to make this without getting carpal tunnel because each tile took at most three clicks to set.
![image](https://user-images.githubusercontent.com/3204033/80306510-a1e22880-8778-11ea-8711-0d2066b03746.png)
Secondary radial menu

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
tweak: Light tiles now use a radial menu for color selection.
tweak: Light tiles need multitools to be adjusted
imageadd: Light tiles can now display Orange
imageadd: Light tiles have a secondary light cycle config
imageadd: Mapper light tiles have also been updated and had their cycling times adjusted
imagedel: Light tiles no longer cycle into white
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

RaveRadbury committed:
Pulls white out of color cycles

Makes white default
Sets other color cycling tile types to ROYGBIV
Adjusts timings of other color cycling tiles

RaveRadbury committed:
Further adjust color cycling

RaveRadbury commented:
On further investigation, light tile customization is an excellent candidate for radial menus.

Incoming.

RaveRadbury committed:
Makes color changing a radial menu

Changes default tile color to blue
Adds "black" light tile

RaveRadbury commented:
This is neat, I'm going to need some assistance figuring out how to configure `custom_check =` for a multitool, but everything else is great.

Rohesie commented on code/game/turfs/open/floor/light_floor.dm:
```diff
@@ -1,17 +1,17 @@
 	var/can_modify_colour = TRUE
 	tiled_dirt = FALSE
-
+	var/list/lighttile_designs = list()
```
```suggestion
	var/static/list/lighttile_designs = list()
```
This way there's a single list shared around.

Rohesie commented on code/game/turfs/open/floor/light_floor.dm:
```diff
@@ -20,6 +20,19 @@
 /turf/open/floor/light/Initialize()
 	. = ..()
 	update_icon()
+	lighttile_designs = list(
```
```suggestion
	if(!length(lighttile_designs))
		lighttile_designs = list(
```

RaveRadbury committed:
Apply suggestions from code review

Thanks Rohesie!

Co-authored-by: Rohesie <rohesie@gmail.com>

RaveRadbury committed:
Continues adjustments based on Rohesie's review.

Rohesie commented on code/game/turfs/open/floor/light_floor.dm:
```diff
@@ -1,25 +1,43 @@
 	var/can_modify_colour = TRUE
 	tiled_dirt = FALSE
-
+	var/static/list/lighttile_designs = list()
```
```suggestion
	var/static/list/lighttile_designs
```
Small detail.

Rohesie commented on code/game/turfs/open/floor/light_floor.dm:
```diff
@@ -1,25 +1,43 @@
 	var/state = 0//0 = fine, 1 = flickering, 2 = breaking, 3 = broken
-	var/list/coloredlights = list("g", "r", "y", "b", "p", "w", "s","o","g")
-	var/currentcolor = 1
+	var/list/coloredlights = list("r", "o", "y", "g", "b", "i", "v", "w", "s", "z")
```
```suggestion
	var/static/list/coloredlights = list("r", "o", "y", "g", "b", "i", "v", "w", "s", "z")
```
This one doesn't change either. No need to have one unique list per object.

Rohesie commented on code/game/turfs/open/floor/light_floor.dm:
```diff
@@ -52,21 +70,18 @@
-		on = FALSE
-	update_icon()
+/turf/open/floor/light/attackby(obj/item/W, mob/user, params)
+	if(W.tool_behaviour == TOOL_MULTITOOL)
+		. = ..()
+		if(.)
+			return
+		if(!can_modify_colour)
+			return
+		var/choice = show_radial_menu(user,src, lighttile_designs, custom_check = FALSE, radius = 36, require_near = TRUE)
+		if(!choice)
+			return FALSE
+		currentcolor = choice
+		update_icon()
```
Put this in `/turf/open/floor/light/multitool_act(mob/living/user)`

RaveRadbury replied:
Done!

RaveRadbury committed:
Apply suggestions from code review

Co-authored-by: Rohesie <rohesie@gmail.com>

RaveRadbury committed:
Acts on requested changes

Un-Statics colored lights because of dancefloor_a and dancefloor_b

RaveRadbury commented:
@Rohesie I had to un-static `coloredlights` because dancefloor_a and dancefloor_b make use of it as well. What would you recommend I do here?

Rohesie commented:
Leave it unset, as `var/list/coloredlights`
Make a global list with those values, something like `GLOB.default_colored_lights`
Call this proc on `Initialize()`:
```cpp
/turf/open/floor/light/proc/set_coloredlights()
  coloredlights = GLOB.default_colored_lights
```
Then you redefine this proc for the other tiles, which may use global lists of their own.

RaveRadbury commented:
> Leave it unset, as `var/list/coloredlights`
> Make a global list with those values, something like `GLOB.default_colored_lights`
> Call this proc on `Initialize()`:
> 
> ```c++
> /turf/open/floor/light/proc/set_coloredlights()
>   coloredlights = GLOB.default_colored_lights
> ```
> 
> Then you redefine this proc for the other tiles, which may use global lists of their own.

The other two entries are using it to call a specific icon state rather than a series of states.

Should I just make a singular GLOB for the normal tiles then?

Rohesie commented:
Yep!

Rohesie commented on code/game/turfs/open/floor/light_floor.dm:
```diff
@@ -70,18 +70,17 @@
+		return FALSE
+	currentcolor = choice
+	update_icon()
 
```
```suggestion
	if(!choice)
		return
	currentcolor = choice
	update_icon()
	return TRUE
```
If you don't return `TRUE` at the end of this proc it will continue with the attack.

RaveRadbury committed:
Adds custom_check callback

Restatics coloredlights,
Modifies mapper light tiles to use currentcolor rather than coloredlights
